### ISSUE TITLE
🤖 Implement ML Training Pipeline Workflow

### ISSUE THREAD

**Issue by norrisaftcc** [norrisaftcc/tool-algoflow-hs#14]
## Overview
Create a machine learning training pipeline that demonstrates AlgoFlow's ability to handle complex, iterative processes with dependencies and resource management.

## Context
- Use Free Monad DSL for maximum flexibility and introspection
- Show integration with ML libraries (hmatrix, tensorflow-haskell)
- Demonstrate checkpointing and resumability
- Handle GPU/CPU resource allocation

### Pipeline Components

1. **Data Preparation**
   - [ ] Load training/validation/test splits
   - [ ] Feature engineering pipeline
   - [ ] Data augmentation (for images/text)
   - [ ] Batch generation with proper shuffling

2. **Model Definition**
   - [ ] Type-safe model architecture specification
   - [ ] Hyperparameter configuration
   - [ ] Model versioning and tracking
   - [ ] Support multiple model types (neural nets, trees, etc.)

3. **Training Loop**
   - [ ] Epoch-based training with early stopping
   - [ ] Gradient accumulation for large batches
   - [ ] Learning rate scheduling
   - [ ] Metric tracking (loss, accuracy, custom metrics)

4. **Evaluation & Validation**
   - [ ] Cross-validation support
   - [ ] Confusion matrix generation
   - [ ] ROC/PR curve calculation
   - [ ] Statistical significance testing

5. **Model Deployment Prep**
   - [ ] Model serialization
   - [ ] Optimization (quantization, pruning)
   - [ ] Deployment package creation
   - [ ] A/B testing setup

### Implementation Approach

```haskell
-- Using Free Monad for flexibility
mlPipeline :: MLWorkflow ModelArtifact
mlPipeline = do
    -- Data preparation
    dataset <- loadDataset "train.csv"
    features <- engineerFeatures dataset
    
    -- Model setup
    model <- defineModel myArchitecture
    
    -- Training with checkpointing
    trainedModel <- trainWithCheckpoints TrainConfig {
        epochs = 100,
        batchSize = 32,
        learningRate = 0.001,
        checkpointEvery = 10
    } model features
    
    -- Evaluation
    metrics <- evaluate trainedModel validationSet
    
    -- Conditional deployment
    when (accuracy metrics > 0.95) $
        prepareDeployment trainedModel
    
    return $ ModelArtifact trainedModel metrics
```

### Key Features to Demonstrate

- [ ] **Type-safe configurations**: Model architecture can't be misconfigured
- [ ] **Resource management**: Automatic GPU allocation/deallocation
- [ ] **Experiment tracking**: All runs logged with parameters
- [ ] **Reproducibility**: Same inputs → same outputs
- [ ] **Fault tolerance**: Resume from checkpoints after failure
- [ ] **Parallel hyperparameter search**: Multiple models training simultaneously

### Integration Examples

1. **With HMatrix**
   ```haskell
   -- Linear algebra operations
   computeGradient :: Matrix Double -> Vector Double -> Vector Double
   ```

2. **With Accelerate**
   ```haskell
   -- GPU computation
   gpuMatMul :: Acc (Matrix Float) -> Acc (Matrix Float) -> Acc (Matrix Float)
   ```

3. **With TensorFlow Haskell**
   ```haskell
   -- Deep learning models
   buildNeuralNet :: TensorFlow (Model Float)
   ```

## File Structure
```
algoflow-clean/
├── src/
│   └── Flow/
│       └── Examples/
│           ├── MLPipeline.hs           # Main implementation
│           ├── MLPipeline/
│           │   ├── DataLoader.hs       # Data handling
│           │   ├── Features.hs         # Feature engineering
│           │   ├── Models/
│           │   │   ├── Neural.hs       # Neural networks
│           │   │   ├── Trees.hs        # Decision trees
│           │   │   └── Linear.hs       # Linear models
│           │   ├── Training.hs         # Training loops
│           │   ├── Evaluation.hs       # Metrics & evaluation
│           │   └── Deployment.hs       # Deployment prep
│           └── MLPipeline/Types.hs     # Type definitions
└── test/
    ├── data/
    │   └── ml/                         # Sample datasets
    └── Flow/Examples/MLPipelineSpec.hs
```

## Acceptance Criteria

- [ ] Complete ML pipeline from data to deployment
- [ ] Type safety prevents common ML bugs (shape mismatches, etc.)
- [ ] Checkpointing allows resuming long training runs
- [ ] Resource usage is optimized (GPU utilization > 80%)
- [ ] Experiment tracking captures all relevant metadata
- [ ] Documentation includes best practices guide

## Performance Requirements

- Train simple neural net in < 5 minutes
- Checkpoint save/restore < 10 seconds
- Support models up to 100M parameters
- Efficient batch processing (>1000 samples/sec)

## Documentation Requirements

- [ ] Tutorial: "From Pandas to AlgoFlow ML"
- [ ] Architecture diagrams for each pipeline stage
- [ ] Comparison with MLflow/Kubeflow
- [ ] Integration guide for popular ML libraries
- [ ] Troubleshooting guide for common issues

## Related Issues
- #4: Test suite implementation
- #8: Workflow validation framework
- #13: Data processing pipeline (for feature engineering)

## Resources
- HMatrix docs: https://hackage.haskell.org/package/hmatrix
- Accelerate: https://www.acceleratehs.org/
- ML best practices: https://developers.google.com/machine-learning/guides
